Issue using Concat on List of Enum Types

### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python

 with pl.StringCache():
        schema = {
            "id": pl.Int64,
            "subtypes": pl.List(pl.Enum(["A", "B", "C", "D"])),
        }

        df1 = pl.DataFrame(
            {
                "id": [1, 2, 3],
                "subtypes": [["A", "B"], ["B", "C"], ["C", "D"]],
            },
            schema=schema,
        )
        df2 = pl.DataFrame(
            {
                "id": [4, 5, 6],
                "subtypes": [["D", "A"], ["A", "B"], ["B", "C"]],
            },
            schema=schema,
        )

    df = pl.concat([df1, df2])
```

### Log output

```shell
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

items = [shape: (3, 2)
┌─────┬────────────┐
│ id  ┆ subtypes   │
│ --- ┆ ---        │
│ i64 ┆ list[enum] │
╞═════╪════════════... list[enum] │
╞═════╪════════════╡
│ 4   ┆ ["D", "A"] │
│ 5   ┆ ["A", "B"] │
│ 6   ┆ ["B", "C"] │
└─────┴────────────┘]

    def concat(
        items: Iterable[PolarsType],
        *,
        how: ConcatMethod = "vertical",
        rechunk: bool = True,
        parallel: bool = True,
    ) -> PolarsType:
        """
        Combine multiple DataFrames, LazyFrames, or Series into a single object.
    
        Parameters
        ----------
        items
            DataFrames, LazyFrames, or Series to concatenate.
        how : {'vertical', 'vertical_relaxed', 'diagonal', 'diagonal_relaxed', 'horizontal', 'align'}
            Series only support the `vertical` strategy.
    
            * vertical: Applies multiple `vstack` operations.
            * vertical_relaxed: Same as `vertical`, but additionally coerces columns to
              their common supertype *if* they are mismatched (eg: Int32 → Int64).
            * diagonal: Finds a union between the column schemas and fills missing column
              values with `null`.
            * diagonal_relaxed: Same as `diagonal`, but additionally coerces columns to
              their common supertype *if* they are mismatched (eg: Int32 → Int64).
            * horizontal: Stacks Series from DataFrames horizontally and fills with `null`
              if the lengths don't match.
            * align: Combines frames horizontally, auto-determining the common key columns
              and aligning rows using the same logic as `align_frames`; this behaviour is
              patterned after a full outer join, but does not handle column-name collision.
              (If you need more control, you should use a suitable join method instead).
        rechunk
            Make sure that the result data is in contiguous memory.
        parallel
            Only relevant for LazyFrames. This determines if the concatenated
            lazy computations may be executed in parallel.
    
        Examples
        --------
        >>> df1 = pl.DataFrame({"a": [1], "b": [3]})
        >>> df2 = pl.DataFrame({"a": [2], "b": [4]})
        >>> pl.concat([df1, df2])  # default is 'vertical' strategy
        shape: (2, 2)
        ┌─────┬─────┐
        │ a   ┆ b   │
        │ --- ┆ --- │
        │ i64 ┆ i64 │
        ╞═════╪═════╡
        │ 1   ┆ 3   │
        │ 2   ┆ 4   │
        └─────┴─────┘
    
        >>> df1 = pl.DataFrame({"a": [1], "b": [3]})
        >>> df2 = pl.DataFrame({"a": [2.5], "b": [4]})
        >>> pl.concat([df1, df2], how="vertical_relaxed")  # 'a' coerced into f64
        shape: (2, 2)
        ┌─────┬─────┐
        │ a   ┆ b   │
        │ --- ┆ --- │
        │ f64 ┆ i64 │
        ╞═════╪═════╡
        │ 1.0 ┆ 3   │
        │ 2.5 ┆ 4   │
        └─────┴─────┘
    
        >>> df_h1 = pl.DataFrame({"l1": [1, 2], "l2": [3, 4]})
        >>> df_h2 = pl.DataFrame({"r1": [5, 6], "r2": [7, 8], "r3": [9, 10]})
        >>> pl.concat([df_h1, df_h2], how="horizontal")
        shape: (2, 5)
        ┌─────┬─────┬─────┬─────┬─────┐
        │ l1  ┆ l2  ┆ r1  ┆ r2  ┆ r3  │
        │ --- ┆ --- ┆ --- ┆ --- ┆ --- │
        │ i64 ┆ i64 ┆ i64 ┆ i64 ┆ i64 │
        ╞═════╪═════╪═════╪═════╪═════╡
        │ 1   ┆ 3   ┆ 5   ┆ 7   ┆ 9   │
        │ 2   ┆ 4   ┆ 6   ┆ 8   ┆ 10  │
        └─────┴─────┴─────┴─────┴─────┘
    
        >>> df_d1 = pl.DataFrame({"a": [1], "b": [3]})
        >>> df_d2 = pl.DataFrame({"a": [2], "c": [4]})
        >>> pl.concat([df_d1, df_d2], how="diagonal")
        shape: (2, 3)
        ┌─────┬──────┬──────┐
        │ a   ┆ b    ┆ c    │
        │ --- ┆ ---  ┆ ---  │
        │ i64 ┆ i64  ┆ i64  │
        ╞═════╪══════╪══════╡
        │ 1   ┆ 3    ┆ null │
        │ 2   ┆ null ┆ 4    │
        └─────┴──────┴──────┘
    
        >>> df_a1 = pl.DataFrame({"id": [1, 2], "x": [3, 4]})
        >>> df_a2 = pl.DataFrame({"id": [2, 3], "y": [5, 6]})
        >>> df_a3 = pl.DataFrame({"id": [1, 3], "z": [7, 8]})
        >>> pl.concat([df_a1, df_a2, df_a3], how="align")
        shape: (3, 4)
        ┌─────┬──────┬──────┬──────┐
        │ id  ┆ x    ┆ y    ┆ z    │
        │ --- ┆ ---  ┆ ---  ┆ ---  │
        │ i64 ┆ i64  ┆ i64  ┆ i64  │
        ╞═════╪══════╪══════╪══════╡
        │ 1   ┆ 3    ┆ null ┆ 7    │
        │ 2   ┆ 4    ┆ 5    ┆ null │
        │ 3   ┆ null ┆ 6    ┆ 8    │
        └─────┴──────┴──────┴──────┘
    
        """  # noqa: W505
        # unpack/standardise (handles generator input)
        elems = list(items)
    
        if not len(elems) > 0:
            raise ValueError("cannot concat empty list")
        elif len(elems) == 1 and isinstance(
            elems[0], (pl.DataFrame, pl.Series, pl.LazyFrame)
        ):
            return elems[0]
    
        if how == "align":
            if not isinstance(elems[0], (pl.DataFrame, pl.LazyFrame)):
                raise TypeError(
                    f"'align' strategy is not supported for {type(elems[0]).__name__!r}"
                )
    
            # establish common columns, maintaining the order in which they appear
            all_columns = list(chain.from_iterable(e.columns for e in elems))
            key = {v: k for k, v in enumerate(ordered_unique(all_columns))}
            common_cols = sorted(
                reduce(
                    lambda x, y: set(x) & set(y),  # type: ignore[arg-type, return-value]
                    chain(e.columns for e in elems),
                ),
                key=lambda k: key.get(k, 0),
            )
            # align the frame data using an outer join with no suffix-resolution
            # (so we raise an error in case of column collision, like "horizontal")
            lf: LazyFrame = reduce(
                lambda x, y: (
                    x.join(y, how="outer", on=common_cols, suffix="_PL_CONCAT_RIGHT")
                    # Coalesce outer join columns
                    .with_columns(
                        [
                            F.coalesce([name, f"{name}_PL_CONCAT_RIGHT"])
                            for name in common_cols
                        ]
                    )
                    .drop([f"{name}_PL_CONCAT_RIGHT" for name in common_cols])
                ),
                [df.lazy() for df in elems],
            ).sort(by=common_cols)
    
            eager = isinstance(elems[0], pl.DataFrame)
            return lf.collect() if eager else lf  # type: ignore[return-value]
    
        out: Series | DataFrame | LazyFrame | Expr
        first = elems[0]
    
        if isinstance(first, pl.DataFrame):
            if how == "vertical":
>               out = wrap_df(plr.concat_df(elems))
E               polars.exceptions.StringCacheMismatchError: cannot compare categoricals coming from different sources, consider setting a global StringCache.
E               
E               Help: if you're using Python, this may look something like:
E               
E                   with pl.StringCache():
E                       # Initialize Categoricals.
E                       df1 = pl.DataFrame({'a': ['1', '2']}, schema={'a': pl.Categorical})
E                       df2 = pl.DataFrame({'a': ['1', '3']}, schema={'a': pl.Categorical})
E                   # Your operations go here.
E                   pl.concat([df1, df2])
E               
E               Alternatively, if the performance cost is acceptable, you could just set:
E               
E                   import polars as pl
E                   pl.enable_string_cache()
E               
E               on startup.
```


### Issue description

Unable to call `pl.concat` on dataframe with Lists of Enum as columns. Suggestion from error is to put code under string cache but that doesn't seem to solve the issue.

### Expected behavior

```
pl.concat([df1.cast({"subtypes": pl.List(pl.Utf8)}), df2.cast({"subtypes": pl.List(pl.Utf8)})]).cast(schema)
``` 
works correctly and gives the desired result.:
```
shape: (6, 2)
┌─────┬────────────┐
│ id  ┆ subtypes   │
│ --- ┆ ---        │
│ i64 ┆ list[enum] │
╞═════╪════════════╡
│ 1   ┆ ["A", "B"] │
│ 2   ┆ ["B", "C"] │
│ 3   ┆ ["C", "D"] │
│ 4   ┆ ["D", "A"] │
│ 5   ┆ ["A", "B"] │
│ 6   ┆ ["B", "C"] │
└─────┴────────────┘
```

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.6-rc.1
Index type:           UInt32
Platform:             macOS-13.6.3-x86_64-i386-64bit
Python:               3.11.4 (v3.11.4:d2340ef257, Jun  6 2023, 19:15:51) [Clang 13.0.0 (clang-1300.0.29.30)]
----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            0.15.1
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           <not installed>
numpy:                1.26.2
openpyxl:             <not installed>
pandas:               <not installed>
pyarrow:              15.0.0
pydantic:             2.5.3
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.23
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

--- a/crates/polars-core/src/datatypes/dtype.rs
+++ b/crates/polars-core/src/datatypes/dtype.rs
@@ -450,7 +450,10 @@ pub fn merge_dtypes(left: &DataType, right: &DataType) -> PolarsResult<DataType>
                     merger.merge_map(rev_map_r)?;
                     Categorical(Some(merger.finish()), *ordering)
                 },
-                (RevMapping::Local(_, idl), RevMapping::Local(_, idr)) if idl == idr => {
+                (RevMapping::Local(_, idl), RevMapping::Local(_, idr))
+                | (RevMapping::Enum(_, idl), RevMapping::Enum(_, idr))
+                    if idl == idr =>
+                {
                     left.clone()
                 },
                 _ => polars_bail!(string_cache_mismatch),

--- a/py-polars/tests/unit/datatypes/test_enum.py
+++ b/py-polars/tests/unit/datatypes/test_enum.py
@@ -70,6 +70,24 @@ def test_nested_enum_creation() -> None:
     assert s.dtype == dtype
 
 
+def test_nested_enum_concat() -> None:
+    dtype = pl.List(pl.Enum(["a", "b", "c", "d"]))
+    s1 = pl.Series([[None, "a"], ["b", "c"]], dtype=dtype)
+    s2 = pl.Series([["c", "d"], ["a", None]], dtype=dtype)
+    expected = pl.Series(
+        [
+            [None, "a"],
+            ["b", "c"],
+            ["c", "d"],
+            ["a", None],
+        ],
+        dtype=dtype,
+    )
+
+    assert_series_equal(pl.concat((s1, s2)), expected)
+    assert_series_equal(s1.extend(s2), expected)
+
+
 def test_casting_to_an_enum_from_utf() -> None:
     dtype = pl.Enum(["a", "b", "c"])
     s = pl.Series([None, "a", "b", "c"])

